src/wayland/input_method/mod.rs

```diff
@@ -0,0 +1,219 @@
+//!
+//! // Add the seat state to your state and create manager globals
+//! InputMethodManagerState::new::<State>(&display_handle);
+//! // Add text input capabilities, needed for the input method to work
+//! TextInputManagerState::new::<State>(&display_handle);
+//!
+//! // create the seat
+//! let seat = Seat::<State>::new(
+//!     &display_handle,          // the display
+//!     "seat-0",                 // the name of the seat, will be advertized to clients
+//!     None                      // insert a logger here
+//! );
+//!
+//! seat.add_input_method(XkbConfig::default(), 200, 25);
+//! // Add input method capabilities to a seat
+//!
+//! ```
+//! ### Run usage
+//!
+//! Once the input method and text input cabailities have been added to a seat,
+//! use the [`seat.input_method().set_point`] function to set the top left point
+//! of a focused surface. This is used to calculate the popup surface location.
+//!
+//! Then use the [`seat.input_method().coordinates`] and [`seat.input_method().with_surface`]
+//! functions to draw the popup surface.
+//!
+
+use wayland_server::{backend::GlobalId, Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New};
+
+use wayland_protocols_misc::zwp_input_method_v2::server::{
+    zwp_input_method_manager_v2::{self, ZwpInputMethodManagerV2},
+    zwp_input_method_v2::ZwpInputMethodV2,
+};
+
+use crate::wayland::seat::Seat;
+
+pub use input_method_handle::{InputMethodHandle, InputMethodUserData};
+pub use input_method_keyboard_grab::InputMethodKeyboardUserData;
+pub use input_method_popup_surface::InputMethodPopupSurfaceUserData;
+
+use super::{seat::XkbConfig, text_input::TextInputHandle};
+
+const MANAGER_VERSION: u32 = 1;
+
+mod input_method_handle;
+mod input_method_keyboard_grab;
+mod input_method_popup_surface;
+
+/// Extends [Seat] with input method functionality
+pub trait InputMethodSeat {
+    /// Add an input method to this seat, and configures the associated keyboard.
+    /// Input methods need different keyboard languages for different input methods.
+    /// E.g a pinyin user will want to use their native keyboard layout, but a
+    /// zhuyin user will always want a taiwanese keyboard layout.
+    fn add_input_method(&self, xkb_config: XkbConfig<'_>, repeat_delay: i32, repeat_rate: i32);
+
+    /// Get an input method associated with this seat
+    fn input_method(&self) -> Option<&InputMethodHandle>;
+}
+
+impl<D: 'static> InputMethodSeat for Seat<D> {
+    fn add_input_method(&self, xkb_config: XkbConfig<'_>, repeat_delay: i32, repeat_rate: i32) {
+        let user_data = self.user_data();
+        user_data.insert_if_missing(InputMethodHandle::default);
+        let input_method = user_data.get::<InputMethodHandle>().unwrap().clone();
+        input_method.configure_keyboard(xkb_config, repeat_delay, repeat_rate);
+    }
+
+    fn input_method(&self) -> Option<&InputMethodHandle> {
+        let user_data = self.user_data();
+        user_data.get::<InputMethodHandle>()
+    }
```

Comment:
Just making sure that I understand correctly, we return Option here instead of inserting `InputMethodHandle::default` because input method needs to have the keyboard configured first using `add_input_method`?

If the keyboard configuration requirement is not there, I imagine that it could look like this:
```rs
trait InputMethodSeat {
   fn input_method(&self) -> &InputMethodHandle;
}

impl InputMethodSeat for Seat {
   fn input_method(&self) -> &InputMethodHandle {
        let user_data = self.user_data();
        user_data.insert_if_missing(InputMethodHandle::default);
        user_data.get::<InputMethodHandle>().unwrap()
   }
}

let seat: Seat = ...;
seat.input_method().configure_keyboard(...);
seat.input_method().watever_else(...);
```

Reply:
Yes, that is correct, but I will try your suggestion and see how that turns out.

Reply:
I figured it was best to leave it as is